### PR TITLE
Improve parameter symbol implementation

### PR DESCRIPTION
**Pull request by pubudu91**
## Purpose
This PR makes `ParameterSymbol` a proper symbol by extending from `Symbol`. This also resolves a bug we had where if the parameters were looked up directly, the returned symbols were var symbols.

## Remarks
I introduced separate flags for the different types of params since the other solution of introducing a separate internal symbol for the params turned out to be a bigger task than expected.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/AbstractCompletionProvider.java [hunk 1]
```diff
@@ -164,6 +167,13 @@ public void sort(BallerinaCompletionContext context, T node, List<LSCompletionIt
                 CompletionItem variableCItem = VariableCompletionItemBuilder.build(varSymbol, varSymbol.getName().get(),
                         typeName);
                 completionItems.add(new SymbolCompletionItem(ctx, symbol, variableCItem));
+            } else if (symbol.kind() == PARAMETER) {
+                ParameterSymbol varSymbol = (ParameterSymbol) symbol;
+                TypeSymbol typeDesc = (varSymbol).typeDescriptor();
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/AbstractCompletionProvider.java:172]: unnecessary parenthesis

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/AbstractCompletionProvider.java [hunk 2]
```diff
@@ -164,6 +167,13 @@ public void sort(BallerinaCompletionContext context, T node, List<LSCompletionIt
...
+                String typeName = typeDesc == null ? "" : CommonUtil.getModifiedTypeName(ctx, typeDesc);
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/AbstractCompletionProvider.java:173]: here the type desc could be null, instead what if we make the typeDescriptor API to return an Optional, which is to be consistent with the other APIs

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/AbstractCompletionProvider.java (hunk 2 above)
Comment (pubudu91) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/AbstractCompletionProvider.java]: Removed this null check. There shouldn't be a case where the typedesc becomes null.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/ParameterCompletionItemBuilder.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/ParameterCompletionItemBuilder.java:2]: Year 2021?

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by rdhananjaya — APPROVED** [review]

**Comment by MaryamZi** [comment]
Can't we get rid of the the `defaultableParam` field in `BVarSymbol` with this change? https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BVarSymbol.java#L40

**Review by MaryamZi — APPROVED** [review]

**Comment by pubudu91** [comment]
> Can't we get rid of the the `defaultableParam` field in `BVarSymbol` with this change? https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/BVarSymbol.java#L40

Should be able to. Will check and address separately.

**Pull request merged**